Lib/asyncio/locks.py

```diff
@@ -105,19 +108,16 @@ async def acquire(self):
         fut = self._get_loop().create_future()
         self._waiters.append(fut)
 
-        # Finally block should be called before the CancelledError
-        # handling as we don't want CancelledError to call
-        # _wake_up_first() and attempt to wake up itself.
         try:
-            try:
-                await fut
-            finally:
-                self._waiters.remove(fut)
-        except exceptions.CancelledError:
+            await fut
+        except BaseException:
+            self._waiters.remove(fut)
             if not self._locked:
+                # Error occurred after release() was called, must re-do release.
```

Comment:
Thank you for your thoughts.  Yes, the comment was too concise and your version is spot on.

I go over the self._waiters in a comment below.  Here is how I follow the logic:

1.  `_wake_up_first()` always wakes up the same Task, at the head of the queue.
2. Only when a task resumes execution, does it remove itself from the queue, opening the possbility for a different Task to run.  When it resumes execution, it _either_ claims the lock for itself, _or_ gives the task at the _new_ head of the queue the chance to claim it.
3. When a task isn't about to claim the lock, and `self._locked()` is False, it is _always_ safe to call `_wake_up_first()`, even multiple times.

Looking at the case you mention:
1. two waiters are cancelled.  They still are in the `self._waiters` queue.
2. first one wakes up.  Removes itself from `self._waiters`. Decides it doesn't want the lock, sees that `self._locked` is False (no one else has it), and so calls `self._wake_up_first().  This is a no-op, since the future has already been cancelled.
3. Second one wakes up.  Removes itself from `self._waiters`.  Decides it doesn't want the lock, seelf that `self._locked` is False, calls `self._wake_up_first`.
4. If there was a Task at the head, then it is now awoken.

Not sure how to formulate that into an invariant, but something like this:

```python
if not self._locked and not claiming_lock, and self._waiters:
  # ensure that the head of the queue will run
  head = self.waiters[0]
  head.set_result(True)
```

In other words, if the lock is not claimed, then the head of the queue, if any, should no longer be blocked.